pyccel/parser/semantic.py

```diff
@@ -2165,14 +2256,15 @@ def _visit_FunctionalFor(self, expr, **settings):
         d_var['shape'] = PythonTuple(*shape)
         d_var['is_stack_array'] = False # PythonTuples can be stack arrays
 
-        lhs_name = _get_name(expr.lhs)
-
-        lhs_empty = NumpyEmpty(shape, dtype = dtype)
-        lhs       = self._create_variable(lhs_name, dtype, lhs_empty, d_var)
+        # ...
+        # TODO [YG, 30.10.2020]:
+        #  - Check if we should use self._create_variable or self._assign_lhs_variable
+        #  - Check if we should allow the possibility that is_stack_array=True
+        #  - Verify that the LHS variable did not exist already
+        lhs = Variable(dtype, _get_name(expr.lhs), **d_var)
         self.insert_variable(lhs)
-        lhs_assign = Assign(lhs, lhs_empty)
-
-        lhs = self.get_variable(lhs_name)
+        lhs_alloc = Allocate(lhs, shape=shape, order=d_var.get('order', 'C'), status='unallocated')
```

Comment:
We don't, I suppose.

But I don't think that our implementation of `FunctionalFor` is correct anyway: instead of `Assign(lhs, FunctionalFor)` we just have `FunctionalFor`.

At some point we should test `FunctionalFor` extensively and fix what is needed, but that's beyond the scope of this PR.

Reply:
I agree, but if it's equivalent to `Assign(lhs, FunctionalFor)` then there's nothing stopping it being in a for loop. That is why I was suggesting calling `_assign_lhs_variable` here

Reply:
Did you?

Reply:
I think I talked about it in our zoom call. So there's probably no trace. But if we do that then the problem should be handled shouldn't it?

Reply:
I cannot quite figure out how `_assign_lhs_variable` should be used in this case. It has no docstring! I tried
```python
ne = []
lhs = self._assign_lhs_variable(expr.lhs, d_var, rhs=expr, new_expressions=ne, is_augassign=False, **settings)
lhs_alloc = ne[0]
```
but no `Allocation` node is created in this case because `expr.lhs.allocatable` is `False`. Any (working) suggestion?

Reply:
Hmm, interesting. This looks right to me. `expr.lhs.allocatable` is `False`, but `d_var['allocatable']` is true so the new `lhs` should be allocatable. Did you make sure you removed the call to `_create_variable` too when you tested this?

Reply:
I don't see any call to `_create_variable` to be removed. I just changed the lines
```python
lhs = Variable(dtype, _get_name(expr.lhs), **d_var)
self.insert_variable(lhs)
lhs_alloc = Allocate(lhs, shape=shape, order=d_var.get('order', 'C'), status='unallocated')
```
with
```python
ne = []
lhs = self._assign_lhs_variable(expr.lhs, d_var, rhs=expr, new_expressions=ne, is_augassign=False, **settings)
lhs_alloc = ne[0]
```
In `_assign_lhs_variable` I check `lhs.allocatable`, which worked in all cases so far. Without extensive documentation of function `_assign_lhs_variable`, I am basically working on a trial and error basis. Clearly this has become unsustainable now.

I think you are more familiar than me with the semantics module, so feel free to try and make changes to this branch.

Reply:
I have fixed it. It wasn't working because `expr.lhs` was `a[i]` not `a`